virtio-devices/src/vhost_user/fs.rs

```diff
@@ -452,35 +438,63 @@ impl VirtioDevice for Fs {
             None
         };
 
-        let mut handler = VhostUserEpollHandler::new(VhostUserEpollConfig {
+        // The backend acknowledged features must contain the protocol feature
+        // bit in case it was initially set but lost through the features
+        // negotiation with the guest.
+        let backend_acked_features = self.common.acked_features
+            | (self.common.avail_features & VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits());
+
+        setup_vhost_user(
+            &mut self.vu.lock().unwrap(),
+            &mem.memory(),
+            queues.clone(),
+            queue_evts.iter().map(|q| q.try_clone().unwrap()).collect(),
+            &interrupt_cb,
+            backend_acked_features,
+            &slave_req_handler,
+        )
+        .map_err(ActivateError::VhostUserFsSetup)?;
+
+        // Run a dedicated thread for handling potential reconnections with
+        // the backend as well as requests initiated by the backend.
+        let (kill_evt, pause_evt) = self.common.dup_eventfds();
+        let mut reconnect_handler: VhostUserEpollHandler<SlaveReqHandler> = VhostUserEpollHandler {
+            vu: self.vu.clone(),
+            mem,
             kill_evt,
             pause_evt,
+            queues,
+            queue_evts,
+            virtio_interrupt: interrupt_cb,
+            acked_features: backend_acked_features,
+            acked_protocol_features: self.acked_protocol_features,
+            socket_path: self.socket_path.clone(),
+            server: false,
             slave_req_handler,
-        });
+        };
 
         let paused = self.common.paused.clone();
         let paused_sync = self.common.paused_sync.clone();
-        let mut epoll_threads = Vec::new();
+
         let virtio_vhost_fs_seccomp_filter =
             get_seccomp_filter(&self.seccomp_action, Thread::VirtioVhostFs)
```

Comment:
When performing test with virtiofsd, I found some problem here:

As this `VhostUserEpollHandler` is also used for handling reconnection, if we will keep the seccomp rules of `VirtioVhostFs`, we need to extent the rule list in `virtio-devices/src/seccomp_filters.rs`:
```rust
fn virtio_vhost_fs_thread_rules() -> Vec<SyscallRuleSet> {
    vec![
        // ....
        allow_syscall(libc::SYS_socket),
        allow_syscall(libc::SYS_nanosleep),
        allow_syscall(libc::SYS_connect),
        allow_syscall(libc::SYS_sendto),
    ]
}
``` 
The reconnection works fine after adding these 4 syscalls to the whitelist.

Reply:
Nice! Let me add these and we should be good to go :)

Reply:
Done. Please give this one more try, and if you confirm this is all good, we can go ahead and merge it :)

Reply:
Just tried, works fine. 😊

Reply:
Thank you @zhangjaycee , I really appreciate you validating this :)